### PR TITLE
Add/keycloak bootstrap

### PR DESCRIPTION
**Pull request by Kidswiss**
## Summary

* Short summary of what's included in the PR
* Give special note to breaking changes

## Checklist

- [ ] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog
- [ ] Update tests.
- [ ] Link this PR to related issues.

### CODE REVIEW

File: pkg/comp-functions/functions/vshnkeycloak/deploy.go [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+	strictHostname := false
+	if comp.Spec.Parameters.Service.FQDN != "" {
+		fqdn = comp.Spec.Parameters.Service.FQDN
+		strictHostname = true
+	}
+
+	extraEnv, err := toYaml([]map[string]any{
+		{
+			"name": "JAVA_OPTS",
+			"value": `-XX:+UseContainerSupport
```
Comment (megian) [inline pkg/comp-functions/functions/vshnkeycloak/deploy.go:212]: Not required anymore https://github.com/projectsyn/component-keycloak/pull/260

File: pkg/comp-functions/functions/vshnkeycloak/deploy.go [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+-XX:MaxRAMPercentage=50.0
+-Djava.net.preferIPv4Stack=true
```
Comment (megian) [inline pkg/comp-functions/functions/vshnkeycloak/deploy.go:214]: Most probably not required anymore >= v22. https://github.com/projectsyn/component-keycloak/issues/231

File: pkg/comp-functions/functions/vshnkeycloak/deploy.go [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+	if comp.Spec.Parameters.Service.FQDN != "" {
+		fqdn = comp.Spec.Parameters.Service.FQDN
+		strictHostname = true
+	}
+
+	extraEnv, err := toYaml([]map[string]any{
+		{
+			"name": "JAVA_OPTS",
+			"value": `-XX:+UseContainerSupport
+-XX:MaxRAMPercentage=50.0
```
Comment (megian) [inline pkg/comp-functions/functions/vshnkeycloak/deploy.go:213]: I think this is broken. The defaults are pretty well. But I haven't yet an other solution. Taking the defaults probably makes more sense.

File: pkg/comp-functions/functions/vshnkeycloak/deploy.go [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+-Djava.net.preferIPv4Stack=true
+-Djava.awt.headless=true
```
Comment (megian) [inline pkg/comp-functions/functions/vshnkeycloak/deploy.go:215]: Not required anymore. https://github.com/projectsyn/component-keycloak/pull/262

File: pkg/comp-functions/functions/vshnkeycloak/deploy.go [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+		{
+			"name":  "KEYCLOAK_ADMIN",
+			"value": "admin",
+		},
+		{
+			"name":  "KC_HOSTNAME",
+			"value": fqdn,
+		},
+		{
+			"name":  "KC_HOSTNAME_STRICT",
```
Comment (megian) [inline pkg/comp-functions/functions/vshnkeycloak/deploy.go:236]: This is for HTTP only. HTTPS is strict by default.

File: pkg/comp-functions/functions/vshnkeycloak/deploy.go [hunk 6]
```diff
@@ -0,0 +1,373 @@
...
+				"memory": res.ReqMem,
+				"cpu":    res.ReqCPU,
+			},
+			"limits": map[string]any{
+				"memory": res.Mem,
+				"cpu":    res.CPU,
+			},
+		},
+		"nodeSelector":      nodeSelector,
+		"args":              []string{"start", "--http-enabled=true"},
```
Comment (megian) [inline pkg/comp-functions/functions/vshnkeycloak/deploy.go:308]: I don't think HTTP should be enabled for new setups. The keycloakx helm chart is somehow security broken. https://github.com/projectsyn/component-keycloak/issues/142

File: apis/vshn/v1/dbaas_vshn_keycloak.go [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+// VSHNKeycloakServiceSpec contains keycloak DBaaS specific properties
+type VSHNKeycloakServiceSpec struct {
+	// FQDN contains the FQDN which will be used for the ingress.
+	// If it's not set, no ingress will be deployed.
+	// This also enables strict hostname checking for this FQDN.
+	FQDN string `json:"fqdn,omitempty"`
+
+	// RelativePath on which Keycloak will listen.
+	// +kubebuilder:default="/"
+	RelativePath string `json:"relativePath,omitempty"`
```
Comment (megian) [inline apis/vshn/v1/dbaas_vshn_keycloak.go:85]: Should be / and not /auth for new installations. `/auth` was legacy for Wildfly.

File: pkg/comp-functions/functions/vshnkeycloak/deploy.go [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+	}
+
+	extraEnv, err := toYaml([]map[string]any{
+		{
+			"name": "JAVA_OPTS",
+			"value": `-XX:+UseContainerSupport
+-XX:MaxRAMPercentage=50.0
+-Djava.net.preferIPv4Stack=true
+-Djava.awt.headless=true
+-Djgroups.dns.query={{ include "keycloak.fullname" . }}-headless`,
```
Comment (megian) [inline pkg/comp-functions/functions/vshnkeycloak/deploy.go:216]: The keycloak operator uses the variable (not JDK parameter):
```
    - name: jgroups.dns.query
      value: "{{ include "keycloak.fullname" . }}-headless"
```

File: pkg/comp-functions/functions/vshnkeycloak/deploy.go [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	startupProbe, err := toYaml(map[string]any{
+		"httpGet": map[string]string{
+			"path": `{{ tpl .Values.http.relativePath $ | trimSuffix "/" }}/health`,
```
Comment (megian) [inline pkg/comp-functions/functions/vshnkeycloak/deploy.go:279]:
```suggestion
			"path": `{{ tpl .Values.http.relativePath $ | trimSuffix "/" }}/health/live`,
```

**Review by megian — CHANGES_REQUESTED** [review]

File: pkg/comp-functions/functions/vshnkeycloak/deploy.go (hunk 6 above)
Comment (Kidswiss) [reply pkg/comp-functions/functions/vshnkeycloak/deploy.go:308]: TLS and all it's headaches will be part of another story. Here we're implementing the basic bootstrapping.

File: apis/vshn/v1/dbaas_vshn_keycloak.go (hunk 7 above)
Comment (Kidswiss) [reply apis/vshn/v1/dbaas_vshn_keycloak.go:85]: `/` is the default if nothing is specified.

File: pkg/comp-functions/functions/vshnkeycloak/deploy.go (hunk 6 above)
Comment (megian) [reply pkg/comp-functions/functions/vshnkeycloak/deploy.go:308]: Kubernetes itself doesn't validate the TLS certificate. So health checks and liveness probes work out of the box. But also here the keycloakx helm chart - with the sting handling - does make it more complicated than it should be https://github.com/projectsyn/component-keycloak/pull/182 as reference.

File: pkg/comp-functions/functions/vshnkeycloak/deploy.go [hunk 10]
```diff
@@ -0,0 +1,373 @@
...
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	values = map[string]any{
+		"replicas": "1",
+		"image": map[string]any{
+			"tag": "23.0.2",
```
[the remark this answers is not in the document]
Comment (Kidswiss) [reply pkg/comp-functions/functions/vshnkeycloak/deploy.go:295]: Version handling will be done via [the helm maintenance](https://github.com/vshn/appcat/blob/master/pkg/maintenance/helm/helmpatcher.go#L39) later.

File: apis/vshn/v1/dbaas_vshn_keycloak.go (hunk 7 above)
Comment (megian) [reply apis/vshn/v1/dbaas_vshn_keycloak.go:85]: Well unfortunately the helm chart keycloakx has `/auth`as default. https://github.com/codecentric/helm-charts/blob/master/charts/keycloakx/values.yaml#L419

File: pkg/comp-functions/functions/vshnkeycloak/deploy.go (hunk 5 above)
Comment (Kidswiss) [reply pkg/comp-functions/functions/vshnkeycloak/deploy.go:236]: I currently keep it, because if no name is specified then keycloak refuses to start entirely.

File: apis/vshn/v1/dbaas_vshn_keycloak.go (hunk 7 above)
Comment (Kidswiss) [reply apis/vshn/v1/dbaas_vshn_keycloak.go:85]: `// +kubebuilder:default="/"` sets the default when the CR is created.

**Review by wejdross — APPROVED** [review]
quite big MR, I have just few notes here, it must be well tested in lab

File: apis/stackgres/v1/sgpoolconfigs.go [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+// SGPoolingConfig is the API for creating pgbouncer configs clusters.
+type SGPoolingConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of a VSHNPostgreSQL.
+	Spec SGPoolingConfigSpec `json:"spec"`
+
+	// Status reflects the observed state of a VSHNPostgreSQL.
+	Status SGPoolingConfigStatus `json:"status,omitempty"`
```
[the remark this answers is not in the document]
Comment (Kidswiss) [reply apis/stackgres/v1/sgpoolconfigs.go:18]: Good catch, I'll change the postgresql references.

File: Makefile [hunk 12]
```diff
@@ -91,6 +91,12 @@ generate-stackgres-crds:
 	go run github.com/deepmap/oapi-codegen/cmd/oapi-codegen --package=v1 -generate=types -o apis/stackgres/v1/sgcluster.gen.go apis/stackgres/v1/sgcluster.yaml
 	perl -i -0pe 's/\*struct\s\{\n\s\sAdditionalProperties\smap\[string\]string\s`json:"-"`\n\s}/map\[string\]string/gms' apis/stackgres/v1/sgcluster.gen.go
 
+	# curl ${STACKGRES_CRD_URL}/SGPoolingConfig.yaml?inline=false -o apis/stackgres/v1/sgpoolconfigs_crd.yaml
+	# yq -i e apis/stackgres/v1/sgpoolconfigs.yaml --expression ".components.schemas.SGPoolingConfigSpec=load(\"apis/stackgres/v1/sgpoolconfigs_crd.yaml\").spec.versions[0].schema.openAPIV3Schema.properties.spec"
+	# yq -i e apis/stackgres/v1/sgpoolconfigs.yaml --expression ".components.schemas.SGPoolingConfigStatus=load(\"apis/stackgres/v1/sgpoolconfigs_crd.yaml\").spec.versions[0].schema.openAPIV3Schema.properties.status"
+	# go run github.com/deepmap/oapi-codegen/cmd/oapi-codegen --package=v1 -generate=types -o apis/stackgres/v1/sgpoolconfigs.gen.go apis/stackgres/v1/sgpoolconfigs.yaml
+	# perl -i -0pe 's/\*struct\s\{\n\s\sAdditionalProperties\smap\[string\]string\s`json:"-"`\n\s}/map\[string\]string/gms' apis/stackgres/v1/sgpoolconfigs.gen.go
+
```
[the remark this answers is not in the document]
Comment (Kidswiss) [reply Makefile:99]: For future us. It probably makes sense if I add a comment though.

tl;dr: the generator for the pool configs doesn't work very well, as it relies on `map[string]string` which the go tooling can't handle. I guess it works with java though. So I had to replace parts of the generated code.
